File: src/epyfilter/c2b2/posterior_checks.R
check_param_in_ci <- function(dt, dd, last_epi=FALSE, percentile = 95, colname=NULL) {
  quantiles <- c((1 - percentile/100)/2, 1 - (1 - percentile/100)/2)
  if (dd == "last") {
    last_day <- max(dt[window==20, day])
    percentile_dt <- dt[day==last_day, as.list(quantile(.SD, quantiles, na.rm=TRUE)),  .SDcols=paste0("sample", 1:300), by=.(window)]
    truth_dt <- dt[day==last_day, .(window, rt)]
    percentile_dt <- merge(percentile_dt, truth_dt, by=c("window"))
    percentile_dt[, in_ci_last_day := `2.5%` <= rt & rt <= `97.5%`]
    return(percentile_dt[, .(window, in_ci_last_day)])
  } else {
    if (!is.numeric(dd)) {
      stop("dd must be numeric")
    }
    if (max(dt[window==max(dt$window), day]) < dd) {
      # day is tooo late
      dd <- max(dt[window==max(dt$window), day])
    }
    percentile_dt <- dt[day==dd, as.list(quantile(.SD, quantiles, na.rm=TRUE)),  .SDcols=paste0("sample", 1:300), by=.(window)]
    truth_dt <- dt[day==dd, .(window, rt)]
    percentile_dt <- merge(percentile_dt, truth_dt, by=c("window"))
    if (last_epi) {
      percentile_dt[, in_ci := `2.5%` <= rt & rt <= `97.5%`]
      names(percentile_dt)[names(percentile_dt) == 'in_ci'] <- colname
      return(percentile_dt[, .SD, .SDcols = c("window", colname)])
    } else {
      percentile_dt[, in_ci := `2.5%` <= rt & rt <= `97.5%`]
      names(percentile_dt)[names(percentile_dt) == 'in_ci'] <- colname
      return(percentile_dt[, .SD, .SDcols = c("window", colname)])
    }
  }
}

compute_ens_var <- function(dt, dd, last_epi=FALSE, colname=NULL) {
  samplecols <- paste0("sample", 1:300)
  if (dd == "last") {
    last_day <- max(dt[window==20, day])
    return(dt[day==last_day, .(ens_var_last_day=rowVars(as.matrix(.SD), na.rm=TRUE)),
              .SDcols=samplecols, by=window])
  } else {
    if (!is.numeric(dd)) {
      stop("day must be numeric")
    }
    if (nrow(dt[day==dd]) < 20) {
      # day is tooo late
      dd <- max(dt[window==20, day])
    }
    return_dt <- dt[day==dd, rowVars(as.matrix(.SD), na.rm=TRUE), .SDcols=samplecols, by=window]
    setnames(return_dt, "V1", colname)
    return(return_dt)
  }
}

bin_data <- function(d, d_pp, num_bins, bins = NULL) {
  data_min <- min(c(min(d), min(d_pp)))
  data_max <- max(c(max(d), max(d_pp)))
  
  if (is.null(bins)) {
    bins <- seq(data_min, data_max, length.out = num_bins)
  }
  
  digitized <- findInterval(d, bins)
  digitized_pp <- findInterval(d_pp, bins)
  
  return(list(digitized, digitized_pp, bins))
}

compute_probs <- function(digitized_data, num_bins) {
  counts <- table(factor(digitized_data, levels = 1:num_bins, exclude = NULL))
  probs <- counts / length(digitized_data)
  return(probs)
}

kl_divergence <- function(p_sample, q_sample, epsilon=1e-10, num_bins=10) {
  stopifnot(length(p_sample) == length(q_sample))
  
  bin_data_result <- bin_data(p_sample, q_sample, num_bins)
  p_bins <- bin_data_result[[1]]
  q_bins <- bin_data_result[[2]]
  bins <- bin_data_result[[3]]
  
  p_probs <- compute_probs(p_bins, num_bins)
  q_probs <- compute_probs(q_bins, num_bins)
  
  p_safe = p_probs + epsilon
  q_safe = q_probs + epsilon
  
  return(sum(p_safe * log(p_safe / q_safe)))
}

avg_kl_divergence <- function(
    dt,
    synthetic_dt,
    data_dt,
    i_ppc,
    dd = NULL,
    evaluate_on = FALSE,
    min_i = 100,
    num_bins = 10,
    colname=NULL) {
  days <- synthetic_dt[i >= min_i, day]
  
  if (!is.null(dd)) {
    dd <- min(dd, max(dt[window==20, day]))
    days <- days[days <= dd]
  }
  
  if (evaluate_on) {
    days = c(dd)
  }
  
  samplecols <- paste0("sample", 1:300)
  
  avg_kl_window <- function(w) {
    tmp_list <- c()
    for (t in days) {
      # Calculate kl
      kl <- kl_divergence(data_dt[day == t, ..samplecols],
                         i_ppc[day == t & window == w, ..samplecols],
                         num_bins = num_bins)
      tmp_list <- c(tmp_list, kl)
    }
    return(data.table(window = w, avg_kl = mean(tmp_list)))
  }
  
  dt_return <- rbindlist(lapply(unique(dt$window), avg_kl_window))
  
  names(dt_return)[names(dt_return) == 'avg_kl'] <- colname
  return(dt_return)
}


wasserstein2 <- function(p_sample, q_sample, num_bins = 10) {
  stopifnot(length(p_sample) == length(q_sample))
  
  bin_data_result <- bin_data(p_sample, q_sample, num_bins)
  p_bins <- bin_data_result[[1]]
  q_bins <- bin_data_result[[2]]
  bins <- bin_data_result[[3]]
  
  p_probs <- compute_probs(p_bins, num_bins)
  q_probs <- compute_probs(q_bins, num_bins)
  
  stopifnot(length(p_probs) == length(q_probs))
  # stopifnot(sum(p_probs) == 1.0, paste("Dist p must sum to 1", sum(p_probs)))
  # stopifnot(sum(q_probs) == 1.0, paste("Dist q must sum to 1", sum(q_probs)))
  
  # Compute the cumulative sums
  p_cdf <- cumsum(p_probs)
  q_cdf <- cumsum(q_probs)
  
  # Calculate the squared distances between the cumulative sums
  squared_distances <- (p_cdf - q_cdf)^2
  
  # Compute the W-2 metric
  w2 <- sqrt(sum(squared_distances))
  
  return(w2)
}


avg_wasserstein2 <- function(dt, synthetic_dt, data_dt, i_ppc, dd=NULL, evaluate_on=FALSE, min_i = 100, num_bins = 10, colname=NULL) {
  days <- synthetic_dt[i >= min_i, day]
  if (!is.null(dd)) {
    dd <- min(dd, max(dt[window==20, day]))
    days <- days[days <= dd]
  }
  
  if (evaluate_on) {
    days = c(dd)
  }
  
  samplecols <- paste0("sample", 1:300)
  
  avg_w2_window <- function(w) {
    tmp_list <- c()
    for (t in days) {
      # Calculate w2
      w2 <- wasserstein2(data_dt[day == t, ..samplecols],
                         i_ppc[day == t & window == w, ..samplecols],
                         num_bins = num_bins)
      tmp_list <- c(tmp_list, w2)
    }
    return(data.table(window = w, avg_w2 = mean(tmp_list)))
  }
  dt_return <- rbindlist(lapply(unique(dt$window), avg_w2_window))
  
  names(dt_return)[names(dt_return) == 'avg_w2'] <- colname
  return(dt_return)
}

free_sim <- function(beta_dt, synthetic_dt) {
  m <- 300
  N <- 100000
  S0 <- 99900
  I0 <- 100
  
  samplecols <- paste0("sample", 1:300)
  free_sim_w <- function(w) {
    S <- matrix(rep(S0,m), nrow = 1, ncol = m)
    Ir <- matrix(rep(I0,m), nrow = 1, ncol = m)
    R <- matrix(rep(0, m), nrow = 1, ncol = m)
    i <- matrix(rep(0, m), nrow = 1, ncol = m)
    
    for (t in 1:max(beta_dt[window==w, day])) {
      if (t < 10) {
        beta <- synthetic_dt[day == t, rt] * gamma
        dSI <- rpois(m, beta * Ir[t,] * S[t,] / N)
      } else {
        beta <- as.matrix(beta_dt[window == w & day == t, ..samplecols])
        dSI <- rpois(m, beta * Ir[t,] * S[t,] / N)
      }
      dIR <- rpois(m, Ir[t,] * gamma)
      
      S_new <- pmin(pmax(S[t, ] - dSI, 0), N)
      I_new <- pmin(pmax(Ir[t, ] + dSI - dIR, 0), N)
      R_new <- pmin(pmax(R[t, ] + dIR, 0), N)
      
      S <- rbind(S, S_new)
      Ir <- rbind(Ir, I_new)
      R <- rbind(R, R_new)
      i <- rbind(i, dSI)
    }
    tmp_dt <- data.table(i)
    colnames(tmp_dt) <- samplecols
    tmp_dt$day <- 0:max(beta_dt[window==w, day])
    tmp_dt$window <- w
    
    return(tmp_dt)
  }
  
  dt_list <- lapply(unique(beta_dt$window), free_sim_w)
  i_ppc <- rbindlist(dt_list)
  
  return(i_ppc)
}

data_rmse <- function(dt, synthetic_dt, dd=NULL, last_epi=FALSE, colname=NULL) {
  # generate i from dt
  synthetic_dt$day <- 1:nrow(synthetic_dt)
  samplecols <- paste0("sample", 1:300)
  beta_dt <- dt[, ..samplecols] * gamma
  beta_dt$day <- dt$day
  beta_dt$window <- dt$window
  i_ppc <- free_sim(beta_dt, synthetic_dt)
  i_dt <- merge(i_ppc, synthetic_dt[, .(day,i)], by="day")

  # compute rmse
  i_dt$sqdiff <- rowMeans((i_dt$i - i_dt[, samplecols, with=FALSE])^2)
  
  if (last_epi) {
    result_dt <- i_dt[day<=dd, sqrt(mean(sqdiff, na.rm=TRUE)), by=window]
    setnames(result_dt, "V1", colname)
    return(result_dt)
  }
  
  return(list(rmse_dt = i_dt[, .(data_rmse=sqrt(mean(sqdiff, na.rm=TRUE))), by=window], i_ppc = i_ppc))
}

rt_rmse <- function(dt, dd=NULL, evaluate_on=FALSE, colname=NULL) {
  dt_copy <- copy(dt)
  samplecols <- paste0("sample", 1:300)
  dt_copy$sqdiff <- rowMeans((dt_copy$rt - dt_copy[, samplecols, with=FALSE])^2)
  
  if (is.null(dd)) {
    # average across entire time series
    return(dt_copy[, .(colname=sqrt(mean(sqdiff, na.rm=TRUE))), by=window])
  }
  
  # day dd should be smaller than the last eval in the window
  dd <- min(dd, max(dt_copy[window==20, day]))
  
  if (evaluate_on) {
    # compute rmse on dd
    result_dt <- dt_copy[day == dd, sqrt(mean(sqdiff, na.rm=TRUE)), by=window]
    setnames(result_dt, "V1", colname)
    return(result_dt)
  }
  else {
    # compute rmse from day 1 until dd
    result_dt <- dt_copy[day <= dd, sqrt(mean(sqdiff, na.rm=TRUE)), by=window]
    setnames(result_dt, "V1", colname)
    return(result_dt)
  }
}

compute_crps <- function(dt, dd, colname, ww=20) {
  # Calculate the Continuous Ranked Probability Score (CRPS).
  # Evaluates on day.
  
  # Parameters:
  #   dt (data.table): The EpiEstim posterior samples
  #   dd (int): The day for which to compute the crps
  #   colname (string): column name for output data table
  #   ww (int): The window to check for the day and obs cutoff
  
  # Returns:
  #   crps_dt (data.table): The CRPS score by window
  
  dt_copy <- copy(dt)
  # Handle day out of bounds
  dd <- min(dd, max(dt_copy[window==ww, day]))
  
  # get observation and ensemble
  samplecols <- paste0("sample", 1:300)
  ensembles <- dt_copy[day==dd, samplecols, with=FALSE]
  ensembles <- as.numeric(ensembles)
  observation <- dt_copy[day==dd & window==ww, i]
  
  hist_result <- hist(ensembles, breaks = length(unique(ensembles)), plot = FALSE)
  bins <- hist_result$breaks
  hist <- hist_result$counts
  cdf <- cumsum(hist / length(ensembles))
  
  heaviside <- function(x) {
    if (x > 0) {
      return(1)
    } else if (x == 0) {
      return(0.5)
    } else {
      return(0)
    }
  }
  heaviside <- Vectorize(heaviside)
  
  crps_scores <- (cdf - heaviside(bins - observation)[2:length(bins)])^2 * diff(bins)
  # CRPS scores must be non-negative
  stopifnot(all(crps_scores >= 0.0)) 
  crps_score <- sum(crps_scores)
  
  crps_dt <- data.table(window=unique(dt_copy$window), crps=crps_score)
  setnames(crps_dt, "crps", colname)
  
  return(crps_dt)
}

File: src/epykalman/utils/posterior_checks.py
import numpy as np


def check_param_in_ci(kf, day, percentile=95):
    """
    Check if the parameter estimate falls within the credible interval.
    Evaluates on day.

    Args:
        kf (object): The EnsembleAdjustmentKalmanFilter object.
        day (int or str): The day for which to check the parameter estimate.
                          If "last", checks the last day.
        percentile (float, optional): The percentile for credible interval. Default is 95.

    Returns:
        bool: True if the parameter estimate is within the credible interval, False otherwise.
    """
    post_betas = np.asarray([θ.beta * θ.t_I for θ in kf.θ_list])
    quantiles = [(1 - percentile / 100) / 2, 1 - (1 - percentile / 100) / 2]
    quantiles_beta = np.quantile(post_betas, q=quantiles, axis=1)
    lower = quantiles_beta[0, :]
    upper = quantiles_beta[1, :]

    if day == "last":
        return lower[-1] <= kf.data.beta[-1] * kf.data.t_I <= upper[-1]
    else:
        assert isinstance(day, int), "day must be integer"
        if day >= len(kf.data.beta):
            # day is past the end of the time series
            return lower[-1] <= kf.data.beta[-1] * kf.data.t_I <= upper[-1]
        else:
            return lower[day] <= kf.data.beta[day] * kf.data.t_I <= upper[day]


def crps(kf, day):
    """
    Calculate the Continuous Ranked Probability Score (CRPS).
    Evaluates on day.

    Parameters:
        kf (object): The EnsembleAdjustmentKalmanFilter object.
        day (int or str): The day for which to compute the ensemble variance.

    Returns:
        crps_score (float): The CRPS score.
    """
    ensembles = np.array([x.i for x in kf.x_list])
    if day >= ensembles.shape[0]:
        day = -1
    ensembles = np.sort(ensembles[day,:])
    obs = kf.data.i[day]
    m = len(ensembles)
    if obs < ensembles[0]:
        cdf_obs = np.ones(m)
        cdf = np.linspace(0,(m-1)/m,m)
        all_mem = np.insert(ensembles,0,obs)  # prepend obs
        delta_fc= np.diff(all_mem)
    elif obs > ensembles[-1]:
        cdf_obs = np.zeros(m)
        cdf = np.linspace(1/m,1,m)
        all_mem = np.append(ensembles,obs)  # append obs
        delta_fc= np.diff(all_mem)
    elif obs in ensembles:
        cdf_obs = ensembles >= obs
        cdf = np.linspace(1/m,1,m)
        all_mem = ensembles
        delta_fc= np.append(np.diff(all_mem), 0)
    else:
        cdf_obs = np.heaviside(ensembles - obs, 0.5)
        cdf = np.linspace(1/m,(m-1)/m,m-1)
        idx = np.where(cdf_obs == 1.)[0][0]
        cdf = np.insert(cdf, idx, cdf[idx-1])
        all_mem = np.sort(np.append(ensembles, obs))
        delta_fc = np.diff(all_mem)
    crps = np.sum(np.array((cdf - cdf_obs) ** 2)*delta_fc)
    return crps


def compute_ens_var(kf, day):
    """
    Compute the ensemble variance of parameter estimates.
    Evaluates on day.

    Args:
        kf (object): The EnsembleAdjustmentKalmanFilter object.
        day (int or str): The day for which to compute the ensemble variance.
                          If "last", computes for the last day.

    Returns:
        float: The ensemble variance.
    """
    if day == "last":
        return np.var(np.asarray([θ.beta * θ.t_I for θ in kf.θ_list])[-1])
    else:
        assert isinstance(day, int), "day must be integer"
        if day >= len(kf.θ_list):
            # day is past the end of the time series
            return np.var(np.asarray([θ.beta * θ.t_I for θ in kf.θ_list])[-1])
        else:
            return np.var(np.asarray([θ.beta * θ.t_I for θ in kf.θ_list])[day])


def avg_kl_divergence(kf, day=None, evaluate_on=False, min_i=100, num_bins=10):
    """
    Compute the average Kullback-Leibler divergence.
    Evaluates up to day and averages over time.
    Evaluates on day if evaluate_on=True.

    Args:
        kf (object): The EnsembleAdjustmentKalmanFilter object.
        day (int, optional): Day to evaluate up to/ on. Defaults to None.
        evaluate_on(bool, optional): If True, evaluates on day. Defaults to False.
        min_i (int, optional): Minimum number of infected individuals to consider. Defaults to 100.
        num_bins (int, optional): Number of bins for histogram. Defaults to 10.

    Returns:
        float: The average Kullback-Leibler divergence.
    """
    if not hasattr(kf, "i_ppc"):
        betas = np.asarray([θ.beta for θ in kf.θ_list])
        _, _, _, _ = kf.free_sim(betas)

    kl_list = []
    days = np.where(kf.data.i_true >= min_i)[0]
    if day:
        # compute only for days less than the day when the second epidemic ends
        assert isinstance(day, int), "peaks must be integer"
        day = min(day, kf.data.n_t - 1)
        days = days[days <= day]

    if evaluate_on:
        days = [day]

    for t in days:
        # Calculate KL divergence
        kl = kl_divergence(
            kf.data.data_distribution[t, :], kf.i_ppc[t, :], num_bins=num_bins
        )
        kl_list.append(kl)
    return np.mean(kl_list)


def avg_wasserstein2(kf, day=None, evaluate_on=False, min_i=100, num_bins=10):
    """
    Compute the average Wasserstein-2 distance.
    Evaluates up to day and averages over time.
    Evaluates on day if evaluate_on=True.

    Args:
        kf (object): The EnsembleAdjustmentKalmanFilter object.
        day (int, optional): Day to evaluate up to/ on. Defaults to None.
        evaluate_on(bool, optional): If True, evaluates on day. Defaults to False.
        min_i (int, optional): Minimum number of infected individuals to consider. Defaults to 100.
        num_bins (int, optional): Number of bins for histogram. Defaults to 10.

    Returns:
        float: The average Wasserstein-2 distance.
    """
    if not hasattr(kf, "i_ppc"):
        betas = np.asarray([θ.beta for θ in kf.θ_list])
        _, _, _, _ = kf.free_sim(betas)
    w2_list = []
    days = np.where(kf.data.i_true >= min_i)[0]
    if day:
        # compute only for days less than the day when the second epidemic ends
        assert isinstance(day, int), "peaks must be integer"
        day = min(day, kf.data.n_t - 1)
        days = days[days <= day]

    if evaluate_on:
        days = [day]

    for t in days:
        # Calculate w2
        w2 = wasserstein2(
            kf.data.data_distribution[t, :], kf.i_ppc[t, :], num_bins=num_bins
        )
        w2_list.append(w2)
    return np.mean(w2_list)


def avg_kl_divergence_ks(ks, day=None, evaluate_on=False, min_i=100, num_bins=10):
    """
    Compute the average Kullback-Leibler divergence for EnsembleSquareRootSmoother.
    Evaluates up to day and averages over time.
    Evaluates on day if evaluate_on=True.

    Args:
        ks (object): The EnsembleSquareRootSmoother object.
        day (int, optional): Day to evaluate up to/ on. Defaults to None.
        evaluate_on(bool, optional): If True, evaluates on day. Defaults to False.
        min_i (int, optional): Minimum number of infected individuals to consider. Defaults to 100.
        num_bins (int, optional): Number of bins for histogram. Defaults to 10.

    Returns:
        float: The average Kullback-Leibler divergence.
    """
    if not hasattr(ks, "i_ppc"):
        betas = np.asarray([θ.beta for θ in ks.θ_lag_list])
        _, _, _, _ = ks.free_sim(betas)

    kl_list = []
    days = np.where(ks.data.i_true >= min_i)[0]
    days = days[days < len(ks.θ_lag_list)]
    if day:
        # compute only for days less than the day when the second epidemic ends
        assert isinstance(day, int), "peaks must be integer"
        day = min(day, ks.i_ppc.shape[0]-1)
        days = days[days <= day]

    if evaluate_on:
        days = [day]

    for t in days:
        # Calculate KL divergence
        kl = kl_divergence(
            ks.data.data_distribution[t, :], ks.i_ppc[t, :], num_bins=num_bins
        )
        kl_list.append(kl)
    return np.mean(kl_list)


def avg_wasserstein2_ks(ks, day=None, evaluate_on=False, min_i=100, num_bins=10):
    """
    Compute the average Wasserstein-2 distance for EnsembleSquareRootSmoother.
    Evaluates up to day and averages over time.
    Evaluates on day if evaluate_on=True.

    Args:
        ks (object): The EnsembleSquareRootSmoother object.
        day (int, optional): Day to evaluate up to/ on. Defaults to None.
        evaluate_on(bool, optional): If True, evaluates on day. Defaults to False.
        min_i (int, optional): Minimum number of infected individuals to consider. Defaults to 100.
        num_bins (int, optional): Number of bins for histogram. Defaults to 10.

    Returns:
        float: The average Wasserstein-2 distance.
    """

    if not hasattr(ks, "i_ppc"):
        betas = np.asarray([θ.beta for θ in ks.θ_lag_list])
        _, _, _, _ = ks.free_sim(betas)
    w2_list = []
    days = np.where(ks.data.i_true >= min_i)[0]
    days = days[days < len(ks.θ_lag_list)]
    if day:
        # compute only for days less than the day when the second epidemic ends
        assert isinstance(day, int), "peaks must be integer"
        day = min(day, ks.i_ppc.shape[0]-1)
        days = days[days <= day]

    if evaluate_on:
        days = [day]

    for t in days:
        # Calculate w2
        w2 = wasserstein2(
            ks.data.data_distribution[t, :], ks.i_ppc[t, :], num_bins=num_bins
        )
        w2_list.append(w2)
    return np.mean(w2_list)


def compute_ens_var_ks(ks, day):
    """
    Compute the ensemble variance of parameter estimates for EnsembleSquareRootSmoother.
    Evaluates on day.

    Args:
        ks (object): The EnsembleSquareRootSmoother object.
        day (int or str): The day for which to compute the ensemble variance.
                          If "last", computes for the last day.

    Returns:
        float: The ensemble variance.
    """
    if day == "last":
        return np.var(np.asarray([θ.beta * θ.t_I for θ in ks.θ_lag_list])[-1])
    else:
        assert isinstance(day, int), "day must be integer"
        if day > len(ks.θ_lag_list):
            day = len(ks.θ_lag_list) - 1
        return np.var(np.asarray([θ.beta * θ.t_I for θ in ks.θ_lag_list])[day])


def check_param_in_ci_ks(ks, day, percentile=95):
    """
    Check if the parameter estimate falls within the confidence interval for EnsembleSquareRootSmoother.
    Evaluates on day.

    Args:
        ks (object): The EnsembleSquareRootSmoother object.
        day (int or str): The day for which to check the parameter estimate.
                          If "last", checks the last day.
        percentile (float, optional): The percentile for confidence interval. Default is 95.

    Returns:
        bool: True if the parameter estimate is within the confidence interval, False otherwise.
    """
    post_betas = np.asarray([θ.beta * θ.t_I for θ in ks.θ_lag_list])
    quantiles = [(1 - percentile / 100) / 2, 1 - (1 - percentile / 100) / 2]
    quantiles_beta = np.quantile(post_betas, q=quantiles, axis=1)
    lower = quantiles_beta[0, :]
    upper = quantiles_beta[1, :]
    if day == "last":
        return lower[-1] <= ks.data.beta[-1] * ks.data.t_I <= upper[-1]
    else:
        assert isinstance(day, int), "day must be integer"
        if day >= len(ks.θ_lag_list):
            day = len(ks.θ_lag_list) - 1
        return lower[day] <= ks.data.beta[day] * ks.data.t_I <= upper[day]


def bin_data(d, d_pp, num_bins, bins=None):
    data_min = min(np.min(d), np.min(d_pp))
    data_max = max(np.max(d), np.max(d_pp))
    if bins is None:
        bins = np.linspace(data_min, data_max, num_bins + 1)
    digitized = np.digitize(d, bins)
    digitized_pp = np.digitize(d_pp, bins)

    # Adjust bin indices to start from 0
    return digitized - 1, digitized_pp - 1, bins


def compute_probs(digitized_data, num_bins):
    counts = np.bincount(digitized_data, minlength=num_bins)
    probs = counts / len(digitized_data)
    return probs


def kl_divergence(p_sample, q_sample, epsilon=1e-10, num_bins=10):
    p_bins, q_bins, bins = bin_data(p_sample, q_sample, num_bins)

    p_probs = compute_probs(p_bins, num_bins + 1)
    q_probs = compute_probs(q_bins, num_bins + 1)
    assert len(p_probs) == len(q_probs)

    p_safe = p_probs + epsilon
    q_safe = q_probs + epsilon

    return np.sum(p_safe * np.log(p_safe / q_safe))


def wasserstein2(p_sample, q_sample, num_bins=10):
    assert len(p_sample) == len(q_sample)
    p_bins, q_bins, bins = bin_data(p_sample, q_sample, num_bins)
    p_probs = compute_probs(p_bins, num_bins + 1)
    q_probs = compute_probs(q_bins, num_bins + 1)
    assert len(p_probs) == len(q_probs)
    assert np.isclose(np.sum(p_probs), 1.0), "Dist p must sum to 1"
    assert np.isclose(np.sum(q_probs), 1.0), "Dist q must sum to 1"

    # Compute the cumulative sums
    p_cdf = np.cumsum(p_probs)
    q_cdf = np.cumsum(q_probs)

    # Calculate the squared distances between the cumulative sums
    squared_distances = (p_cdf - q_cdf) ** 2

    # Compute the W-2 metric
    w2 = np.sqrt(np.sum(squared_distances))

    return w2


def data_rmse(kf, last_epi_day=None, last_epi=False):
    """
    Compute the root mean square error (RMSE) between the data and case counts posterior.
    Evaluates up to day and averages over time.

    Args:
        kf (object): The EnsembleAdjustmentKalmanFilter object.
        last_epi_day (int, optional): Last day of the second epidemic. Defaults to None.
        last_epi (bool, optional): If True, compute RMSE up to last_epi_day.

    Returns:
        float: The RMSE.
    """
    if last_epi:
        i_kf = np.array([x.i for x in kf.x_list])
        i_kf = i_kf[:last_epi_day]
        rmse = np.sqrt(np.mean((i_kf.T - kf.data.i[1 : len(i_kf) + 1]) ** 2, axis=0))
    else:
        i_kf = np.array([x.i for x in kf.x_list])
        rmse = np.sqrt(np.mean((i_kf.T - kf.data.i[1 : len(i_kf) + 1]) ** 2, axis=0))

    return np.mean(rmse)


def rt_rmse(kf, last_epi=False, peaks=None):
    """
    Compute the root mean square error (RMSE) between the truth and Rt posterior.
    Evaluates up to day and averages over time. If peaks, just evalutes on peak day(s).

    Args:
        kf (object): The EnsembleAdjustmentKalmanFilter object.
        last_epi (bool, optional): If True, compute RMSE up to last_epi_day.
        peaks (int or list of int, optional): Peak day(s) of epidemic(s). Defaults to None.

    Returns:
        float: The RMSE.
    """
    rt_kf = np.array([θ.beta * θ.t_I for θ in kf.θ_list])

    if last_epi:
        # compute rmse from day 1 until the last day of the second epidemic
        assert isinstance(peaks, int), "peaks must be integer"
        peaks = min(peaks, rt_kf.shape[0] - 1)
        rmse = np.sqrt(np.mean((rt_kf[:peaks].T - kf.data.rt[:peaks]) ** 2, axis=0))
        return np.mean(rmse)
    elif peaks is not None:
        # computes rmse on specific day(s)
        if isinstance(peaks, list):
            peaks = [min(p, rt_kf.shape[0] - 1) for p in peaks]
        elif isinstance(peaks, int):
            peaks = min(peaks, rt_kf.shape[0] - 1)
        else:
            raise ValueError("peaks must be a list or an integer")

        rmse = np.sqrt(np.mean((rt_kf[peaks].T - kf.data.rt[peaks]) ** 2, axis=0))
        return np.mean(rmse)
    else:
        # computes rmse for entire time series
        rmse = np.sqrt(np.mean((rt_kf.T - kf.data.rt[: len(rt_kf)]) ** 2, axis=0))
    return np.mean(rmse)
